### PR TITLE
CIP-35: Plutus Core Evolution

### PR DESCRIPTION
**Pull request by michaelpj**
This CIP proposes a process for making CIP proposals to change Plutus
Core or its interface to the ledger.

### CODE REVIEW

File: CIP-0035/README.md [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+Type: Process
+Created: 2022-02-09
+License: CC-BY-4.0
+---
+
+# Evolution of Plutus Core in the Cardano ledger
+
+## Abstract
+
+This CIP proposes a process for proposing changes to Plutus Core itself, and the parts of the Cardano ledger which relate to it.
```
Comment (kozross) [inline CIP-0035/README.md:17]: What specifically are these? Considering Plutus' nontrivial dependency footprint, it would be good to have a list of these, even if it's not 100% exhaustive.

File: CIP-0035/README.md [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+- In the Specification:
+    - Names and types/kinds for the new functions or types.
+    - A source for the implementation (e.g. a library which can be linked against); or a generic description of the functionality which is implementable in any programming language.
+    - In the Rationale:
+    - An argument for the utility of the new builtins.
+
+It SHOULD also include:
+- In the Rationale
+    - Examples of real-world use cases where the new additions would be useful.
+    - A comparison with an implementation using the existing features, and an argument why the builtin is preferable (e.g. better performance).
```
Comment (kozross) [inline CIP-0035/README.md:159]: Who decides that the requirements to move to Draft have been met?

File: CIP-0035/README.md [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+
+In order to move to Proposed status, it MUST include:
+- In the Specification:
+    - For new types: a precise description of the measure used for the size of a value of that type.
+    - For new builtin functions: a costing function for the builtin function.
+- In the Rationale:
+    - If an external implementation is provided: an argument that it is trustworthy.
+    - Discussion of how any measures and costing functions were determined.
+- In the Path to Active section:
+    - A named implementor (person or organization) who will implement the support for the new builtins in the Plutus Core interpreter, or a pull request to the `plutus` repository with an implementation.
```
Comment (kozross) [inline CIP-0035/README.md:169]: Several issues here:

* Again, who decides when the requirements to move to Proposed have been met?
* Based on [here](https://github.com/input-output-hk/plutus/blob/master/plutus-core/cost-model/CostModelGeneration.md) and [here](https://github.com/input-output-hk/plutus/pull/4368#issuecomment-1032025802), the requirement of a costing function being provided is completely unworkable. The only group of people able to cost anything are IOG themselves, and while there is clearly [a goal to extending this](https://github.com/input-output-hk/plutus/issues/4375#issuecomment-1029864939) and [some discussion towards this](https://github.com/input-output-hk/plutus/issues/4384), very non-trivial limitations remain.

File: CIP-0035/README.md [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+
+In order to move to Active status, the following must be true:
+- The external implementations must be available.
+- The `plutus` repository MUST be updated with an implementation including costing.
+- The Plutus Core specification MUST be updated to include the new builtins.
+- The ledger MUST be updated to include new protocol parameters to control costing of the new builtins.
+- The completion of these steps MUST be tracked in the Path to Active section.
```
Comment (kozross) [inline CIP-0035/README.md:176]: All the issues raised in the 'move to Proposed' section apply here as well. In addition:

* No mention of testing. Given issues such as [this](https://github.com/input-output-hk/plutus/pull/4284#issuecomment-996185889), [this](https://github.com/input-output-hk/plutus/pull/4322#discussion_r782817268), and (while not from Plutus, _critically_ relevant) [this](https://github.com/input-output-hk/cardano-base/pull/253), the importance and quality of tests cannot be overstated. We also need to formalize practices such as use of Hedgehog, 1-in-a-thousand as the acceptable failure standard for Hedgehog's bad coverage implementation, and so forth: all the above are described in the links.

**Review by kozross — CHANGES_REQUESTED** [review]
In general, I think it's a good idea to have a more specified and clear structure around modifying Plutus and the libraries it depends on. However, I am concerned at the degree of bureaucracy this adds. Consider, for example, [this issue here](https://github.com/input-output-hk/plutus/issues/4233). Based on the current write-up, I see the process looking like this:

1. CIP for this gets sent. We await its approval, however long that takes, deal with whatever discussions and concerns arise, etc. This is complicated by the fact that we have to modify `cardano-base` as well, which by my reading, requires a _separate_ CIP.
1. Once accepted, the process for getting to Draft must be completed. Some as-yet-unclear party decides when this goal is met, which may require more discussion, dealing with more concerns, etc.
1. Once gotten to Draft, the process for Proposed must be completed. Assuming costing can be done at all (see my note), it's yet another round of the same.
1. Once gotten to Proposed, the process for getting to Active must be completed. You get the idea by now.

This would take something that is already a lengthy and complex process, make it _lengthier_ and _more_ complex, while arguably as-written still failing to solve key issues (for example, see my statement regarding testing). Additionally, a document such as this should document proper practices and give specifics about what needs to be done (again, see my comment regarding testing): at the moment, it's not even clear who gets to decide when we can move, for example, to Draft. Additionally, if my reading is correct, for something like the linked task, this process might have to be completed _twice_.

In short, much more clarity is needed, and the amount of steps specified here need to be considered for usefulness: personally, they seem excessive, and fairly unclear, at this stage.

File: CIP-0035/README.md (hunk 1 above)
Comment (michaelpj) [reply CIP-0035/README.md:17]: Does the list in "Types of change" not answer this question?

File: CIP-0035/README.md (hunk 3 above)
Comment (michaelpj) [reply CIP-0035/README.md:169]: Implementation may be difficult :) But in general the idea isn't meant to be that the implementor has to do everything and the Plutus team just sit back and chill. As with any contribution, if we think we should accept it we're going to help. In the short term that means that the Plutus team will probably do the costing functions.

(See my main comment for why I didn't include more "implementation process" stuff like this in the text.)

File: CIP-0035/README.md (hunk 4 above)
Comment (michaelpj) [reply CIP-0035/README.md:176]: As discussed in my main comment, this is an implementation issue so is out of scope.

**Comment by michaelpj** [comment]
> However, I am concerned at the degree of bureaucracy this adds, and what we actually gain by it.

Partially, the bureaucracy is the point. I believe that the bar for changing the blockchain _should_ be higher. We should be sure that many people in the community want the change, and that the change is well-specified. 

Moreover, the idea is for the development of Cardano to be increasingly driven by people other than IOG, and that includes the design work. It's an easier process for the community to say "we want X" and get IOG to do the design work, but we *want* the community to do the design work!

> Consider, for example, https://github.com/input-output-hk/plutus/issues/4233.

That's a great example. Let me sketch out what I think the process for that would have been:

1. You open a Draft CIP, which is more-or-less like the issue that you opened, but with more details.
2. There is discussion on the CIP. Hopefully, this will surface any difficulties due to the increased number of eyes, for example, it might have identified the fact that you wanted Schnorr signatures.
3. In parallel with that, you can start on the implementation work. The CIP doesn't say anything about when the implementation has to happen. If you can persuade the `cardano-base` maintainers to merge your PR with the implementation etc. then you can go ahead and do that.
4. The remaining details are hammered out in discussion on the CIP and potentially the prototype implementation. Somewhere around here the design is completed enough that the CIP becomes Proposed.
5. The implementation gets reviewed and accepted (or not! making a CIP doesn't guarantee that something will get accepted! it's just an open design doc), and the CIP gets changed to Active.

I think this would have been not that much more onerous than what actually happened. The main differences would be:
- More design work up front for the proposer. This is, as I said, deliberate, but is annoying for you.
- More eyes on the proposal by the community. This is important: part of the point is that we _shouldn't_ just be changing the blockchain because one community member makes a `plutus` issue and I go "seems fine".

> This is complicated by the fact that we have to modify cardano-base as well, which by my reading, requires a separate CIP.

Why do you think that? That wasn't the intention. 

> Some as-yet-unclear party decides when this goal is met

As I commented inline, it's the CIP editors, who are in general responsible for assessing the progression of CIPs.

> Additionally, a document such as this should document proper practices and give specifics about what needs to be done

Unfortunately, this document can't be as precise as I would like, since _implementation_ is out of scope for CIPs. A lot of requirements are buried in "it is the responsibilty of the implementor to get their change accepted", such as testing. I agree that that makes it less helpful than it could be. When I realised this I considered not even making a CIP and just adding this to the `plutus` `CONTRIBUTING` doc or something. But I think it's still valuable to use the CIP process for the _design_ part of the process, even if it has to leave something out.

At least, I'll try and clarify the text to make the implementation requirements more obvious.

File: CIP-0035/README.md (hunk 3 above)
Comment (michaelpj) [reply CIP-0035/README.md:169]: Or perhaps a better way of saying it is: there might be named _implementors_ which could be "MLabs and the Plutus team (for cost modelling)"

File: CIP-0035/README.md (hunk 2 above)
Comment (kozross) [reply CIP-0035/README.md:159]: Thanks for clarifying this: an inline link would help make that clearer in the text.

**Review by L-as — APPROVED** [review]
I think this is fine honestly.

File: CIP-0035/README.md (hunk 1 above)
Comment (michaelpj) [reply CIP-0035/README.md:17]: I guess it's this sentence that's mostly confusing. I thought "This document considers the following types of change: <explicit list of changes>" was pretty explicit but okay.

Comment (Benjmhart) [reply CIP-0035/README.md:17]: i think the direct question here is "My change requires modifications to `cardano-base` as well as the plutus interpreter, does this require two separate CIP procedures?"

File: CIP-0035/README.md (hunk 3 above)
Comment (Benjmhart) [reply CIP-0035/README.md:169]: I'd like to see a process for generating cost reports when an implementation is ready to go,  the point should be to _enable_ the community to do this in the correct way, rather than gating it off to specific implementors.   Teach them to fish.  Ideally the request process would not need to go through a CIP editors meeting, but that may become a standard practice.

File: CIP-0035/README.md (hunk 1 above)
Comment (Benjmhart) [reply CIP-0035/README.md:17]: this seems to be answered 'No' in the below longform comment

Comment (michaelpj) [reply CIP-0035/README.md:17]: What I was trying to clarify is what in the text makes it seem that you *would* need a CIP to modify `cardano-base`, given that "changes to dependencies" is not listed as a type of change covered by the CIP.

**Comment by michaelpj** [comment]
Updated:
- I've attempted to clarify the scope of the CIP a bit more, and that it applies only to the listed types of change.
- I've made it clearer that implementation is out of scope and may introduce more work. I've also removed the requirement that the CIP include named implementors: now it just says that in order to reach Active, it must have been implemented. How that happens (who does it, how they manage it), doesn't need to be specified.

**Comment by rhyslbw** [comment]
CIP Identifier clash with https://github.com/cardano-foundation/CIPs/pull/137

**Comment by michaelpj** [comment]
> CIP Identifier clash with https://github.com/cardano-foundation/CIPs/pull/137

My understanding is that the Editors are responsible for allocating a CIP number anyway: I've only picked one here because otherwise you can't put it in a logical subdirectory :shrug: 

We can change the number of whichever one doesn't get merged first ;)

**Review by kozross — APPROVED** [review]

**Review by KtorZ — APPROVED** [review]

**Review by SebastienGllmt — APPROVED** [review]

**Comment by michaelpj** [comment]
Given the lack of further comments and the approvals from the editors, I've changed the status to Active to represent that fact that we will be using this process going forwards.

**Comment by KtorZ** [comment]
@michaelpj whatever you did, you _might_ want to push it :blush: ?
